### ISSUE TITLE
java.net.SocketException: Too many open files while detecting outliers

### ISSUE THREAD

**Issue by ansell** [AtlasOfLivingAustralia/biocache-store#155]
The Jenkins job detecting outliers is suffering from a large number of exceptions related to having too many files/file handles/sockets open:

```
2016-11-01 05:24:54,475 ERROR: [ExpertDistributionOutlierTool] - Worker(1) experienced error finding distribution outliers for http://id.biodiversity.org.au/node/apni/2893909
java.net.SocketException: Too many open files
	at java.net.Socket.createImpl(Socket.java:447)
	at java.net.Socket.<init>(Socket.java:421)
	at java.net.Socket.<init>(Socket.java:280)
	at org.apache.commons.httpclient.protocol.DefaultProtocolSocketFactory.createSocket(DefaultProtocolSocketFactory.java:80)
	at org.apache.commons.httpclient.protocol.DefaultProtocolSocketFactory.createSocket(DefaultProtocolSocketFactory.java:122)
	at org.apache.commons.httpclient.HttpConnection.open(HttpConnection.java:707)
	at org.apache.commons.httpclient.HttpMethodDirector.executeWithRetry(HttpMethodDirector.java:387)
	at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:171)
	at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:397)
	at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:323)
	at au.org.ala.biocache.tool.ExpertDistributionActor.getExpertDistributionWkt(ExpertDistributionOutlierTool.scala:550)
	at au.org.ala.biocache.tool.ExpertDistributionActor.findOutliersForLsid(ExpertDistributionOutlierTool.scala:317)
	at au.org.ala.biocache.tool.ExpertDistributionActor$$anonfun$act$1$$anonfun$apply$1.applyOrElse(ExpertDistributionOutlierTool.scala:233)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)
	at scala.actors.ReactorTask.run(ReactorTask.scala:31)
	at scala.actors.Reactor$class.resumeReceiver(Reactor.scala:129)
	at au.org.ala.biocache.tool.ExpertDistributionActor.scala$actors$InternalReplyReactor$$super$resumeReceiver(ExpertDistributionOutlierTool.scala:212)
	at scala.actors.InternalReplyReactor$class.resumeReceiver(InternalReplyReactor.scala:59)
	at au.org.ala.biocache.tool.ExpertDistributionActor.resumeReceiver(ExpertDistributionOutlierTool.scala:212)
	at scala.actors.InternalActor$class.searchMailbox(InternalActor.scala:76)
	at au.org.ala.biocache.tool.ExpertDistributionActor.searchMailbox(ExpertDistributionOutlierTool.scala:212)
	at scala.actors.Reactor$$anonfun$startSearch$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(Reactor.scala:117)
	at scala.actors.Reactor$$anonfun$startSearch$1$$anonfun$apply$mcV$sp$1.apply(Reactor.scala:114)
	at scala.actors.Reactor$$anonfun$startSearch$1$$anonfun$apply$mcV$sp$1.apply(Reactor.scala:114)
	at scala.actors.ReactorTask.run(ReactorTask.scala:33)
	at scala.actors.ReactorTask.compute(ReactorTask.scala:63)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by ansell** [comment]
The method in the stacktrace looks like it is responsibly closing sockets:

https://github.com/AtlasOfLivingAustralia/biocache-store/blob/6e8306f8fe5b8152e2841f6f9343e80739ca87d8/src/main/scala/au/org/ala/biocache/tool/ExpertDistributionOutlierTool.scala#L181

However, there are other cases of HttpClient objects that are not closing their sockets which may be contributing to the issue when there are exceptions (a large number of 404 and other exceptions being generated in the log also):

https://github.com/AtlasOfLivingAustralia/biocache-store/blob/master/src/main/scala/au/org/ala/biocache/util/HttpUtil.scala#L16

**Comment by ansell** [comment]
The fix for #156 should have fixed this error, although it only occurs when services are flaky, so difficult to test. Closing this but will reopen if it occurs again in future.

**Comment by ansell** [comment]
Refs https://github.com/AtlasOfLivingAustralia/maintenance/issues/35
